// ribbons.logger/lib/logger.js
var util = require('util');
var colors = require('colors');

colors.setTheme({
  date: 'grey',
  name: 'cyan',
  info: 'magenta',
  warn: 'yellow',
  error: 'red',
  moreDetail: 'green',
  evenMoreDetail: 'blue'
});

function Logger (options) {
  options = options || {};
  this.name = options.name || 'Logger';
  this.debug = options.debug || true;
}
Logger.prototype.info = function() {
  if (this.debug) {
    arguments[0] = arguments[0].toString().info;
    log.apply(this,arguments);
  }
}
Logger.prototype.warn = function() {
  arguments[0] = ('WARNING: '.warn.bold + arguments[0].toString().info);
  log.apply(this,arguments);
}
Logger.prototype.error = function() {
  arguments[0] = ('ERROR: '.error.bold + arguments[0].toString().info);
  log.apply(this,arguments);
}
log = function (/*level, message*/) {
  var args = [].slice.call(arguments);
  console.log(
    String(+new Date()).date + ' ' +
    this.name.name.bold + ' ' +
    args.shift() +
    ((args.length > 0) ? ' ' + args.shift().toString().moreDetail : '') + ' ' +
    args.join(' - ').evenMoreDetail
  );
}

module.exports = Logger;
